### PR TITLE
Feature/fix test prepare raster data by tile

### PR DESCRIPTION
**Pull request by hasruleng**
fix prepare raster data by tile and its unit test

### CODE REVIEW

File: R/ff_prep.R [hunk 1]
```diff
@@ -165,25 +165,34 @@ ff_prep <- function(datafolder = NA, country = NA, shape = NA, tiles = NULL, gro
...
     if (!hasvalue(shape) & (shrink == "extract")) {
       if (!exists("countries", inherits = F)) {
         data(countries, envir = environment())
         borders <- terra::vect(countries)
       }
       shape <- aggregate(intersect(terra::as.polygons(terra::ext(terra::rast(files[1]))), borders))
     }
+    # end-of initialize_shape_from_borders
+
+    # process_tile_dates (process_a_tile_in_dates)
```
Comment (aminbksi) [inline R/ff_prep.R:178]: why this comment?

**Review by aminbksi — CHANGES_REQUESTED** [review]

File: R/ff_prep.R (hunk 1 above)
Comment (hasruleng) [reply R/ff_prep.R:178]: I am trying to fully run ff_prep_refactored. Fyi. it still has an error until now. 
To fix the errors, I need to compare the refactored functions with the original code. With those comments, I can easily see the original code for each refactored functions.

File: R/ff_prep_refactored.R [hunk 2]
```diff
@@ -299,14 +299,19 @@ prepare_raster_data_by_tile <- function(files, shape, shrink, window, verbose) {
...
+  # Return both extent and rasstack as a list
+  return(list(extent = extent, rasstack = rasstack))
 }
 
-load_groundtruth_raster <- function(selected_files, groundtruth_pattern, first, verbose, hasgroundtruth) {
-  #   groundtruth_raster <- NULL
-  #   hasgroundtruth <- FALSE
-
+load_groundtruth_raster <- function(selected_files, groundtruth_pattern, first, verbose, extent, hasgroundtruth) {
+  cat("\n====== load_groundtruth_raster running ======\n")
```
Comment (aminbksi) [inline R/ff_prep_refactored.R:308]: remove the prints

File: R/ff_prep_refactored.R [hunk 3]
```diff
@@ -320,8 +325,14 @@ load_groundtruth_raster <- function(selected_files, groundtruth_pattern, first,
       groundtruth_raster[] <- 0
     }
   }
+  list_gt_raster <- list(groundtruth_raster = groundtruth_raster, hasgroundtruth = hasgroundtruth, first=first)
+  cat("======> return of load_gt_raster =======>\n")
+  print(paste0("groundtruth_raster: ", groundtruth_raster))
```
Comment (aminbksi) [inline R/ff_prep_refactored.R:330]: remove logs

File: R/ff_prep_refactored.R [hunk 4]
```diff
@@ -502,18 +520,23 @@ process_tile_dates <- function(tiles, tile, files, shape, shrink, window, ground
     }
 
     selected_files <- filter_files_by_date_and_groundtruth(date, files, groundtruth_pattern)
-    rasstack <- prepare_raster_data_by_tile(selected_files, shape, shrink, window, verbose)
+    # rasstack <- prepare_raster_data_by_tile(selected_files, shape, shrink, window, verbose)
+    result <- prepare_raster_data_by_tile(files, shape, shrink, window, verbose)
```
Comment (aminbksi) [inline R/ff_prep_refactored.R:524]: why files?

File: R/ff_run.R [hunk 5]
```diff
@@ -147,7 +147,7 @@ ff_run <- function(shape = NULL, country = NULL, prediction_dates = NULL,
       )
       ff_prep_params_combined <- merge_lists(default = ff_prep_params_original, user = ff_prep_params)
       ff_prep_params_combined <- merge_lists(default = ff_prep_params_combined, user = list("inc_features" = fltr_features, "adddate" = F, "addxy" = F, "verbose" = F))
-      traindata <- do.call(ff_prep, ff_prep_params_combined)
+      traindata <- do.call(ff_prep_refactored, ff_prep_params_combined)
```
Comment (aminbksi) [inline R/ff_run.R:150]: only test it locally, dont change other files

File: runner/ff_prep_malloc_error.log [hunk 6]
```diff
@@ -0,0 +1,87 @@
+> # library(ForestForesight)
```
Comment (aminbksi) [inline runner/ff_prep_malloc_error.log:1]: what is this file?

File: runner/run_prediction.R [hunk 7]
```diff
@@ -0,0 +1,60 @@
+# library(ForestForesight)
+code_location <- "C:/Kodingan3/ForestForesight"
+ff_folder <- "C:/Kodingan3/ForestForesight/test_data"
+library("devtools")
+load_all(code_location) # with this we are using the live R code, not the build package of ForestForesight
+library(sf)
+
```
Comment (aminbksi) [inline runner/run_prediction.R:7]: and this file

File: runner/run_prediction_with_ff_prep_refactored.log [hunk 8]
```diff
@@ -0,0 +1,68 @@
+> # library(ForestForesight)
+> code_location <- "C:/Kodingan3/ForestForesight"
+> ff_folder <- "C:/Kodingan3/FFdata"
```
Comment (aminbksi) [inline runner/run_prediction_with_ff_prep_refactored.log:3]: and this

File: runner/run_prediction_with_ori_ff_prep.log [hunk 9]
```diff
@@ -0,0 +1,114 @@
+> # library(ForestForesight)
+> code_location <- "C:/Kodingan3/ForestForesight"
+> ff_folder <- "C:/Kodingan3/FFdata"
```
Comment (aminbksi) [inline runner/run_prediction_with_ori_ff_prep.log:3]: and this file

File: runner/run_train_gabon.R [hunk 10]
```diff
@@ -0,0 +1,59 @@
+
+# Step 1: Set up the parameters
+code_location <- "C:/Kodingan3/ForestForesight/"
+ff_folder <- "C:/Kodingan3/FFdata/"
+country_code <- "GAB"
```
Comment (aminbksi) [inline runner/run_train_gabon.R:5]: and this file?

File: tests/testthat/test-load_groundtruth_raster.R [hunk 11]
```diff
@@ -0,0 +1,38 @@
+library(testthat)
+library(terra)
+
+test_that("load_groundtruth_raster works correctly", {
```
Comment (aminbksi) [inline tests/testthat/test-load_groundtruth_raster.R:4]: does this test pass?

**Review by aminbksi — CHANGES_REQUESTED** [review]

File: R/ff_prep.R (hunk 1 above)
Comment (aminbksi) [reply R/ff_prep.R:178]: but you dont need to merge them to develop, have them locally

File: tests/testthat/test-load_groundtruth_raster.R (hunk 11 above)
Comment (hasruleng) [reply tests/testthat/test-load_groundtruth_raster.R:4]: this one is not finished in this branch

File: R/ff_prep.R (hunk 1 above)
Comment (hasruleng) [reply R/ff_prep.R:178]: you're kinda ignoring the fact that we have to fix ff_prep_refactored.
not only me that need this

File: R/ff_prep_refactored.R (hunk 4 above)
Comment (hasruleng) [reply R/ff_prep_refactored.R]: my bad, I fixed this

File: R/ff_prep_refactored.R [hunk 12]
```diff
@@ -298,15 +298,13 @@ prepare_raster_data_by_tile <- function(files, shape, shrink, window, verbose) {
     cat(paste("with extent", round(extent[1], 5), round(extent[2], 5), round(extent[3], 5), round(extent[4], 5), "\n"))
   }
 
-  rasstack <- terra::rast(sapply(files, function(x) terra::rast(x, win = extent)))
-
-  return(rasstack)
+  rasstack <- terra::rast(sapply(selected_files, function(x) terra::rast(x, win = extent)))
+  # return (rasstack) # Amin's version doesn't return extent variable
```
Comment (aminbksi) [inline R/ff_prep_refactored.R:302]: can remove comments?

**Review by aminbksi — CHANGES_REQUESTED** [review]

**Review by aminbksi — APPROVED** [review]

**Pull request merged**